app/models/commit_status.rb

```diff
@@ -39,33 +46,70 @@ def expire_cache(commit)
 
   def combined_status
     @combined_status ||= begin
-      statuses = [github_status]
+      statuses = [github_state]
       statuses += [release_status, *ref_statuses].compact if @stage
-      statuses[1..-1].each_with_object(statuses[0]) { |status, merged| merge(merged, status) }
+      merge_statuses(statuses)
     end
   end
 
-  def merge(a, b)
-    a[:state] = [a.fetch(:state), b.fetch(:state)].max_by { |state| STATE_PRIORITY.index(state.to_sym) }
-    a.fetch(:statuses).concat b.fetch(:statuses)
-  end
-
+  # Gets a reference's state, combining results from both the Status and Checks API
   # NOTE: reply is an api object that does not support .fetch
-  def github_status
+  def github_state
     static = @reference.match?(Build::SHA1_REGEX) || @reference.match?(Release::VERSION_REGEX)
     expires_in = ->(reply) { cache_duration(reply) }
     cache_fetch_if static, cache_key(@reference), expires_in: expires_in do
-      GITHUB.combined_status(@project.repository_path, @reference).to_h
+      checks_result = with_octokit_client_error_rescue { github_check }
+      status_result = with_octokit_client_error_rescue { github_status }
+
+      merge_statuses([checks_result, status_result])
+    end
+  end
+
+  # Gets commit statuses using GitHub's check API. Currently parsing it to match status structure to better facilitate
+  # transition to new API. See https://developer.github.com/v3/checks/runs/ and
+  # https://developer.github.com/v3/checks/suites/ for details
+  def github_check
+    base_url = "repos/#{@project.repository_path}/commits/#{@reference}"
+    preview_header = {Accept: 'application/vnd.github.antiope-preview+json'}
+
+    check_suites = GITHUB.get("#{base_url}/check-suites", headers: preview_header)[:check_suites]
+    checks = GITHUB.get("#{base_url}/check-runs", headers: preview_header)
+
+    overall_state = check_suites.
+      map { |suite| check_state_equivalent(suite[:conclusion]) }.
+      max_by { |state| STATE_PRIORITY.index(state.to_sym) }
```

Comment:
could add a `|| raise("Unknown: #{state}")` for faster debugging when a new state is added

Reply:
^ this ?

Reply:
Right now we default to 'pending' if there are no `check_suites` to be consistent with the States API..

Reply:
I meant for inside the max_by :D

Reply:
Ah gotcha. Currently returning 'pending' if the `conclusion` is `nil` so I don't have to check both 'status' and 'conclusion'. Added an explicit check for `nil` in `check_state_equivalent` for pending, and a raise if an unexpected state is found.